<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Record-Linkage" data-toc-modified-id="Record-Linkage-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Record Linkage</a></span><ul class="toc-item"><li><span><a href="#The-Principles-of-Record-Linkage" data-toc-modified-id="The-Principles-of-Record-Linkage-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>The Principles of Record Linkage</a></span></li><li><span><a href="#Linking-Patents-to-Grants:-individual-and-organization-levels" data-toc-modified-id="Linking-Patents-to-Grants:-individual-and-organization-levels-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Linking Patents to Grants: individual and organization levels</a></span></li><li><span><a href="#Python-Setup" data-toc-modified-id="Python-Setup-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Python Setup</a></span></li></ul></li><li><span><a href="#Data-Exploration:-Know-Your-Data" data-toc-modified-id="Data-Exploration:-Know-Your-Data-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Data Exploration: Know Your Data</a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Individual-level-links:-inventor-(patent)-<->-principal-investigator-(grant)" data-toc-modified-id="Individual-level-links:-inventor-(patent)-<->-principal-investigator-(grant)-2.0.1"><span class="toc-item-num">2.0.1&nbsp;&nbsp;</span>Individual-level links: inventor (patent) &lt;-&gt; principal investigator (grant)</a></span><ul class="toc-item"><li><span><a href="#PatentsView:-inventor-names-on-patents" data-toc-modified-id="PatentsView:-inventor-names-on-patents-2.0.1.1"><span class="toc-item-num">2.0.1.1&nbsp;&nbsp;</span>PatentsView: inventor names on patents</a></span></li><li><span><a href="#FedRePORTER:-principal-investigator-names-on-grants" data-toc-modified-id="FedRePORTER:-principal-investigator-names-on-grants-2.0.1.2"><span class="toc-item-num">2.0.1.2&nbsp;&nbsp;</span>FedRePORTER: principal investigator names on grants</a></span></li></ul></li></ul></li></ul></li><li><span><a href="#Data-Cleaning" data-toc-modified-id="Data-Cleaning-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Data Cleaning</a></span><ul class="toc-item"><li><span><a href="#The-Importance-of-Pre-Processing" data-toc-modified-id="The-Importance-of-Pre-Processing-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>The Importance of Pre-Processing</a></span><ul class="toc-item"><li><span><a href="#Individual-level-links" data-toc-modified-id="Individual-level-links-3.1.1"><span class="toc-item-num">3.1.1&nbsp;&nbsp;</span>Individual-level links</a></span><ul class="toc-item"><li><span><a href="#Clean-Inventor-Data" data-toc-modified-id="Clean-Inventor-Data-3.1.1.1"><span class="toc-item-num">3.1.1.1&nbsp;&nbsp;</span>Clean Inventor Data</a></span></li><li><span><a href="#Clean-Grant-Data" data-toc-modified-id="Clean-Grant-Data-3.1.1.2"><span class="toc-item-num">3.1.1.2&nbsp;&nbsp;</span>Clean Grant Data</a></span></li></ul></li></ul></li></ul></li><li><span><a href="#Deterministic-Matching" data-toc-modified-id="Deterministic-Matching-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Deterministic Matching</a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Individual-level-links" data-toc-modified-id="Individual-level-links-4.0.1"><span class="toc-item-num">4.0.1&nbsp;&nbsp;</span>Individual-level links</a></span><ul class="toc-item"><li><span><a href="#REGEX-CHEATSHEET" data-toc-modified-id="REGEX-CHEATSHEET-4.0.1.1"><span class="toc-item-num">4.0.1.1&nbsp;&nbsp;</span>REGEX CHEATSHEET</a></span></li><li><span><a href="#EXAMPLES" data-toc-modified-id="EXAMPLES-4.0.1.2"><span class="toc-item-num">4.0.1.2&nbsp;&nbsp;</span>EXAMPLES</a></span></li></ul></li></ul></li><li><span><a href="#References-and-Further-Readings" data-toc-modified-id="References-and-Further-Readings-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>References and Further Readings</a></span><ul class="toc-item"><li><span><a href="#Record-Linkage" data-toc-modified-id="Record-Linkage-4.1.1"><span class="toc-item-num">4.1.1&nbsp;&nbsp;</span>Record Linkage</a></span></li><li><span><a href="#Parsing" data-toc-modified-id="Parsing-4.1.2"><span class="toc-item-num">4.1.2&nbsp;&nbsp;</span>Parsing</a></span></li><li><span><a href="#Regular-Expression" data-toc-modified-id="Regular-Expression-4.1.3"><span class="toc-item-num">4.1.3&nbsp;&nbsp;</span>Regular Expression</a></span></li><li><span><a href="#Record-Linkage-Python-package" data-toc-modified-id="Record-Linkage-Python-package-4.1.4"><span class="toc-item-num">4.1.4&nbsp;&nbsp;</span>Record Linkage Python package</a></span></li></ul></li></ul></li></ul></div>

Contributors:
Rayid Ghani, Frauke Kreuter, Julia Lane, Brian Kim, Adrianne Bradford, Alex Engler, Nicolas Guetta Jeanrenaud, Graham Henke, Daniela Hochfellner, Clayton Hunter, Avishek Kumar, Jonathan Morgan, Ekaterina Levitskaya.

# Record Linkage

This notebook will provide you with an instruction into Record Linkage using Python. Upon completion of this notebook you will be able to apply record linkage techniques using the `recordlinkage` package to combine data from different sources in Python. 
It will lead you through all the steps necessary for a sucessful record linkage starting with data preparation  including pre-processing, cleaning and standardization of data.

## The Principles of Record Linkage
The goal of record linkage is to determine if pairs of records describe the same entity. For instance, this is important for removing duplicates from a data source or joining two separate data sources together. Record linkage also goes by the terms data matching, merge/purge, duplication detection, de-duping, reference matching, entity resolution, disambiguation, co-reference/anaphora in various fields.

There are several approaches to record linkage that include 
    - exact matching,
    - rule-based linking, 
    - probabilistic linking. 
- An example of **exact matching** is joining records based on a direct identifier. This is what we have already have done in SQL by joining tables. 
- **Rule-based matching** involves applying a cascading set of rules that reflect the domain knowledge of the records being linked. 
- In **probabilistic record linkages**, linkage weights are estimated to calculate the probability of a certain match.

In practical applications you will need record linkage techiques to combine information addressing the same entity that is stored in different data sources. Record linkage will also help you to address the quality of different data sources. For example, if one of your databases has missing values you might be able to fill those by finding an identical pair in a different data source. Overall, the main applications of record linkage are:
    1. Merging two or more data files. 
    2. Identifying the intersection of the two data sets. 
    3. Updating data files (with the data row of the other data files) and imputing missing data.
    4. Entity disambiguation and de-duplication.

## Linking Patents to Grants: individual and organization levels

In this class we will link records obtained from the **PatentsView** (https://www.patentsview.org/download/) to the data on federal grants in **FedRePORTER** (https://federalreporter.nih.gov/FileDownload). In both datasets we have names that we can use to link: inventor names on patents and principal investigator names on grants.

**About USPTO/PatentsView**

The PatentsView platform links inventors, organizations, locations, and patenting activity since 1976. 


**Key variables of interest**:

**Patent**
- Patent processing time (duration between a patent application filing date and
grant date)
- Title
- Type of patent document (utility, design, plant, reissue, defensive, and statutory
invention registration—offer different kinds of protection for distinct types of
subject matter)
- Cited patent (patent is cited when it appears as a reference in another patent or
patent application)
- Citing patent (patent that cites a patent or patent application as a reference in its
&#39;references cited&#39; section)
- Grant date (date a patent was issued by the USPTO)
- International Patent Classification (IPC) (system of patent classifications
published by the World Intellectual Property Organization (WIPO))

**Assignee** (the name of the entity—company, foundation, partnership, holding company
or individual—that owns the patent)
- First name (if assignee is individual)
- Last name (if assignee is individual)
- Organization name (if assignee is organization)
- Type (classification of assignee)
- Location (City, State, Country, Latitude, Longitude, County, FIPS code (state), FIPS code (county))

**Inventor** (the individual who conceived the invention)
- First name
- Last name
- Gender
- Location (City, State, Country, Latitude, Longitude, County, FIPS code (state), FIPS code (county))
- Co-inventor (a person who appears on the same patent or patent application as
another inventor)

**About Federal RePORTER**

Federal RePORTER is a collaborative effort led by STAR METRICS® to create a searchable
database of scientific awards from agencies (across agencies or fiscal years, by the award&#39;s
project leader, or by a text search of a project&#39;s title, terms, or abstracts). The National Institutes of Health (NIH) and the National Science Foundation (NSF), under the auspices of Office of Science and Technology Policy (OSTP), are leading this project with funding provided by NIH, NSF, the U.S. Department of Agriculture, and the Environmental Protection Agency.

Agencies with data in Federal RePORTER: DOD (Department of Defense), ED (Department of Education), Environmental Protection Agency (EPA), HHS (Department of Health and Human Services), National Aeronautics and Space Administration (NASA), National Science Foundation (NSF), USDA (United States Department of Agriculture), U.S. Department of Veterans Affairs (VA).

**Key variables of interest**:
- Project description
    - Title
    - Abstract
    - Terms (keywords from the project title and abstract)
- Project duration
    - Start date
    - End date
- Total cost of a project
- Financial year
- Federal agency, department
- Principal Investigator (PI) Project Leader
    - First name
    - Last name
- Other PIs
    - First name
    - Last name
- Organization name
- Organization location (City, State, Zip, Country)
- CFDA code (The Catalog of Federal Domestic Assistance, federal financial assistance
and nonfinancial assistance programs)

**Questions for group discussion:** 
- What variables would you use for record linkage using these datasets?
- What problems do you anticipate?

## Python Setup

**Class Subset**: The FedRePORTER dataset includes records from year 2010, and PatentsView dataset includes records from 2015-2019 of patents with government interest (in order to account for a time lag between the start of a grant and potential outputs (registered patents) from a grant).

Import necessary Python packages.

In [1]:
# data import and manipulation
import pandas as pd

# record linkage and preprocessing
import recordlinkage as rl
from recordlinkage.preprocessing import clean, phonetic

# Data Exploration: Know Your Data

### Individual-level links: inventor (patent) <-> principal investigator (grant)

#### PatentsView: inventor names on patents

In [21]:
# Import CSV into a dataframe
inventors = pd.read_csv('../Data/patentsview_inventors.csv')

In [22]:
# View first rows of the dataframe
inventors.head()

,patent_id,year,inventor_id,name_first,name_last,inventor_city,inventor_state,inventor_country
0,9980635,2018,9980635-1,Maria Eliana,Manquez Hatta,Santiago,NaN,CL
1,9980635,2018,9980635-2,Humberto,Lobos Sucarrat,Santiago,NaN,CL
2,9999665,2018,8398993-2,Pablo A.,Gonzalez Munoz,Santiago,NaN,CL
3,9980641,2018,5321501-6,Joel S.,Schuman,Pittsburgh,PA,US
4,9980641,2018,9980641-3,Lawrence Edward,Kagemann,Pittsburgh,PA,US


In [23]:
inventors.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4841 entries, 0 to 4840
Data columns (total 8 columns):
patent_id           4841 non-null object
year                4841 non-null int64
inventor_id         4841 non-null object
name_first          4841 non-null object
name_last           4841 non-null object
inventor_city       4840 non-null object
inventor_state      3994 non-null object
inventor_country    4841 non-null object
dtypes: int64(1), object(7)
memory usage: 302.6+ KB


We notice that `inventor_state` information appears to be missing in some cases (3,994 records vs total 4,841 records).

We can check the rows that are missing by calling `.isnull()`.

In [24]:
# Show rows which are null in `inventor_state` column
inventors[inventors['inventor_state'].isnull()]

,patent_id,year,inventor_id,name_first,name_last,inventor_city,inventor_state,inventor_country
0,9980635,2018,9980635-1,Maria Eliana,Manquez Hatta,Santiago,NaN,CL
1,9980635,2018,9980635-2,Humberto,Lobos Sucarrat,Santiago,NaN,CL
2,9999665,2018,8398993-2,Pablo A.,Gonzalez Munoz,Santiago,NaN,CL
100,9980645,2018,4306291-3,Dominique M.,Durand,Toronto,NaN,CA
101,9981915,2018,5156079-2,Wei,Li,Toronto,NaN,CA
102,9983412,2018,D419566-1,David Scott,Dunn,Toronto,NaN,CA
103,9984283,2018,7873169-1,Ke,Zeng,Toronto,NaN,CA
104,9989702,2018,7826509-2,Benjamin G.,Lee,Toronto,NaN,CA
105,9993465,2018,5326659-1,Yong,Liu,Toronto,NaN,CA
106,9997705,2018,5546889-4,Yang,Yang,Toronto,NaN,CA


It looks like our records are not filtered by only US.

In [28]:
# Check unique values in the `inventor_country` column
inventors['inventor_country'].unique()
#inventors['inventor_country'].value_counts()

array(['CL', 'US', 'CA', 'FR', 'GB', 'CN', 'JP', 'KR', 'SG', 'DE', 'NL',
       'IN', 'CH', 'AU', 'TW', 'HK', 'FI', 'BE', 'SE', 'SW', 'EN', 'NM',
       'VN', 'BR', 'IE', 'SA', 'ES', 'IT', 'CZ', 'IL', 'RU', 'MX', 'AT',
       'DK', 'MO', 'UY', 'BG', 'PH'], dtype=object)

Let's filter by only those records which are in the US.

In [29]:
# Filter by those rows where country is US
inventors = inventors[inventors['inventor_country'] == 'US']

In [30]:
# Check that now we only have US as an inventor country
inventors['inventor_country'].unique()

array(['US'], dtype=object)

In [55]:
inventors['patent_id'].nunique()

593

We will be linking on name and location. What do you observe in the data?

In [65]:
# Check the tail of the table 
#inventors.tail(10)
#inventors[['name_first','name_last']].value_counts()
inventors.groupby(['name_first','name_last']).size().reset_index().sort_values(0, ascending=[False])

#inventors.loc[(inventors['name_first']=='David C.') & (inventors['name_last']=='Smith'),:]

,name_first,name_last,0
328,David C.,Smith,84
1530,Wei,Huang,74
791,Jun,Li,58
1531,Wei,Li,50
897,Lei,Wang,48
1641,Zhanao,Deng,40
212,Chao,Wang,36
563,Hui,Wang,34
1593,Yang,Liu,31
1595,Yang,Yang,30


- First name includes first and middle names (potentially parse out)
- First and last names are capitalized
- The same person (with the same inventor id) can have different associated locations 

Check most frequent counts of variables.

In [33]:
inventors['inventor_city'].value_counts().head(10)

Cambridge      80
Los Angeles    60
San Diego      57
New York       48
Albuquerque    45
San Jose       44
Palo Alto      43
Baltimore      38
Pasadena       36
Chicago        35
Name: inventor_city, dtype: int64

In [34]:
inventors['inventor_state'].value_counts().head(10)

CA    917
MA    400
NY    258
MD    176
FL    164
IL    163
PA    154
TX    142
NJ    129
CT    122
Name: inventor_state, dtype: int64

#### FedRePORTER: principal investigator names on grants

In [35]:
# Read-in the dataset into a dataframe
pi = pd.read_csv('../Data/fedreporter_pi_names_grants.csv')

In [36]:
# Show first rows of the dataframe
pi.head()

,PROJECT_ID,CONTACT_PI_PROJECT_LEADER,OTHER_PIS,ORGANIZATION_CITY,ORGANIZATION_STATE,ORGANIZATION_COUNTRY
0,121219,"POIRER, DAVID",NaN,TUCSON,AZ,UNITED STATES
1,121220,"HEYWARD, ANN O",NaN,BROOK PARK,OH,UNITED STATES
2,121221,"KALOGERA, VASSILIKI","LUIJTEN, ERIK",EVANSTON,IL,UNITED STATES
3,121222,"GUILDERSON, THOMAS P",NaN,SANTA CRUZ,CA,UNITED STATES
4,121223,"HAYES, GILLIAN","TENTORI, MONICA",IRVINE,CA,UNITED STATES


In [37]:
pi.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 111400 entries, 0 to 111399
Data columns (total 6 columns):
PROJECT_ID                    111400 non-null int64
 CONTACT_PI_PROJECT_LEADER    111398 non-null object
 OTHER_PIS                    10477 non-null object
 ORGANIZATION_CITY            110853 non-null object
 ORGANIZATION_STATE           109933 non-null object
 ORGANIZATION_COUNTRY         110861 non-null object
dtypes: int64(1), object(5)
memory usage: 5.1+ MB


Let's check which countries are there in the dataset.

In [40]:
# Call .unique() function
pi[' ORGANIZATION_COUNTRY'].unique()

array(['UNITED STATES', 'BRAZIL', 'BERMUDA', 'SWITZERLAND', 'FRANCE',
       'CANADA', 'UNITED KINGDOM', 'VENEZUELA', 'SOUTH AFRICA',
       'NEW ZEALAND', 'DENMARK', nan, 'SWEDEN', 'INDIA', 'SPAIN',
       'AUSTRALIA', 'PERU', 'HAITI', 'DOMINICAN REP', 'JAMAICA',
       'NIGERIA', 'ITALY', 'IRELAND', 'BELGIUM', 'CHINA', 'THAILAND',
       'JAPAN', 'ZAMBIA', 'ISRAEL', 'GERMANY', 'NETHERLANDS', 'UGANDA',
       'PAKISTAN', 'KOREA REP OF', 'CHILE', 'TANZANIA U REP', 'ARGENTINA',
       'HUNGARY', 'URUGUAY', 'RUSSIA', 'KENYA', 'MALAWI', 'LITHUANIA',
       'FINLAND', 'MEXICO', 'COLOMBIA', 'CAMBODIA', 'PHILIPPINES',
       'TURKEY', 'ZIMBABWE', 'BOTSWANA', 'CROATIA', 'COSTA RICA',
       'TUNISIA', 'ICELAND', 'MOZAMBIQUE', 'GHANA', 'SINGAPORE',
       'LEBANON', 'PARAGUAY', 'ETHIOPIA', 'UKRAINE', 'LESOTHO',
       "COTE D'IVOIRE", 'DOMINICA', 'NAMIBIA', 'GUATEMALA', 'RWANDA',
       'GEORGIA', 'VIETNAM', 'MOLDOVA', 'TRINIDAD/TOBA', 'CAMEROON',
       'ANGOLA', 'INDONESIA', 'CONGO', 'CONGO 

Let's filter by only US like we did with the PatentsView dataset.

In [41]:
# Filter by only those rows with US as an organization_country
pi = pi[pi[' ORGANIZATION_COUNTRY'] == 'UNITED STATES']

In [42]:
# Call .unique() function
pi[' ORGANIZATION_COUNTRY'].unique()

array(['UNITED STATES'], dtype=object)

In [43]:
pi[' ORGANIZATION_CITY'].value_counts().head(10)

BOSTON          5802
NEW YORK        4603
BEHTESDA        3745
PHILADELPHIA    3044
BALTIMORE       2957
SEATTLE         2917
LOS ANGELES     2905
LA JOLLA        2835
CHICAGO         2523
HOUSTON         2345
Name:  ORGANIZATION_CITY, dtype: int64

In [44]:
pi[' ORGANIZATION_STATE'].value_counts().head(10)

CA    15293
NY     9186
MA     8812
MD     7932
PA     6245
TX     5653
NC     4036
IL     3796
WA     3409
OH     3034
Name:  ORGANIZATION_STATE, dtype: int64

**Compare inventor and grants table: what types of cleaning/standardization would we need to do?**

In [45]:
inventors.head()

,patent_id,year,inventor_id,name_first,name_last,inventor_city,inventor_state,inventor_country
3,9980641,2018,5321501-6,Joel S.,Schuman,Pittsburgh,PA,US
4,9980641,2018,9980641-3,Lawrence Edward,Kagemann,Pittsburgh,PA,US
5,9980645,2018,9037248-2,Brian,Wodlinger,Pittsburgh,PA,US
6,9981048,2018,9981048-1,Ira Jacob,Fox,Pittsburgh,PA,US
7,9981048,2018,9981048-3,Taichiro N.,Nishikawa,Pittsburgh,PA,US


In [46]:
pi.head()

,PROJECT_ID,CONTACT_PI_PROJECT_LEADER,OTHER_PIS,ORGANIZATION_CITY,ORGANIZATION_STATE,ORGANIZATION_COUNTRY
0,121219,"POIRER, DAVID",NaN,TUCSON,AZ,UNITED STATES
1,121220,"HEYWARD, ANN O",NaN,BROOK PARK,OH,UNITED STATES
2,121221,"KALOGERA, VASSILIKI","LUIJTEN, ERIK",EVANSTON,IL,UNITED STATES
3,121222,"GUILDERSON, THOMAS P",NaN,SANTA CRUZ,CA,UNITED STATES
4,121223,"HAYES, GILLIAN","TENTORI, MONICA",IRVINE,CA,UNITED STATES


<span style="color:red">**Checkpoint 1: Explore organization-level data from PatentsView and FedRePORTER**</span>

Now try exploring organization-level data for both PatentsView and FedRePORTER using functions for data exploration used in the examples above.

In [256]:
assignees = pd.read_csv('../Data/patentsview_assignee_org_names.csv')

In [257]:
organizations = pd.read_csv('../Data/fedreporter_org_names_grants.csv')

In [259]:
# your code here...

assignees.loc[assignees.]


(4371, 5)

# Data Cleaning

## The Importance of Pre-Processing
Data pre-processing is an important step in a data anlysis project in general, in record linkage applications in particular. The goal of pre-processing is to transform messy data into a dataset that can be used in a project workflow.

Linking records from different data sources comes with different challenges that need to be addressed by the analyst. The analyst must determine whether or not two entities (individuals, businesses, geographical units) on two different files are the same. This determination is not always easy. In most of the cases there is no common uniquely identifing characteristic for a entity. For example, is Bob Miller from New Yor the same person as Bob Miller from Chicago in a given dataset? This detemination has to be executed carefully because consequences of wrong linkages may be substantial (is person X the same person as the person X on the list of identified terrorists). Pre-processing can help to make better informed decisions.

Pre-processing can be difficult because there are a lot of things to keep in mind. For example, data input errors, such as typos, misspellings, truncation, abbreviations, and missing values need to be corrected. Literature shows that preprocessing can improve matches. 

“In situations of reasonably high-quality data, preprocessing can yield a greater improvement in matching efficiency than string comparators and ‘optimized’ parameters. In some situations, 90% of the improvement in matching efficiency may be due to preprocessing” (Winkler 2009).

The most common reason why matching projects fail is lack of time and resources for data cleaning. 

In the following section we will walk you through some pre-processing steps, these include but are not limited to removing spaces, parsing fields, and standardizing strings.

### Individual-level links

#### Clean Inventor Data
We will start by cleaning and preprocessing the inventor data. We need to remove whitespaces, make sure that everything is in lower case, we need to parse the first name and the last name, and harmonize all the other information we need for the linkage.

The `recordlinkage` package comes with a built-in cleaning function we can use. The `clean()` function removes any characters such as `-`, `.`, `/`, `:`, brackets of all types, and also lowercases by default.

In [68]:
# Let's take a look at names before
inventors.head()

,patent_id,year,inventor_id,name_first,name_last,inventor_city,inventor_state,inventor_country
3,9980641,2018,5321501-6,Joel S.,Schuman,Pittsburgh,PA,US
4,9980641,2018,9980641-3,Lawrence Edward,Kagemann,Pittsburgh,PA,US
5,9980645,2018,9037248-2,Brian,Wodlinger,Pittsburgh,PA,US
6,9981048,2018,9981048-1,Ira Jacob,Fox,Pittsburgh,PA,US
7,9981048,2018,9981048-3,Taichiro N.,Nishikawa,Pittsburgh,PA,US


In [69]:
# Cleaning names (using the record linkage package tool)
inventors['name_last'] = clean(inventors['name_last'])
inventors['name_first'] = clean(inventors['name_first'])

In [70]:
# Let's take a look at names after the cleaning
inventors.head()

,patent_id,year,inventor_id,name_first,name_last,inventor_city,inventor_state,inventor_country
3,9980641,2018,5321501-6,joel s,schuman,Pittsburgh,PA,US
4,9980641,2018,9980641-3,lawrence edward,kagemann,Pittsburgh,PA,US
5,9980645,2018,9037248-2,brian,wodlinger,Pittsburgh,PA,US
6,9981048,2018,9981048-1,ira jacob,fox,Pittsburgh,PA,US
7,9981048,2018,9981048-3,taichiro n,nishikawa,Pittsburgh,PA,US


We can parse and extract middle names.

In [71]:
# Split by white space and extract first element for the first name,
# second element for the middle name
inventors['name_middle'] = inventors.name_first.str.split(' ').str.get(1)
inventors['name_first'] = inventors.name_first.str.split(' ').str.get(0)

In [72]:
inventors.head()

,patent_id,year,inventor_id,name_first,name_last,inventor_city,inventor_state,inventor_country,name_middle
3,9980641,2018,5321501-6,joel,schuman,Pittsburgh,PA,US,s
4,9980641,2018,9980641-3,lawrence,kagemann,Pittsburgh,PA,US,edward
5,9980645,2018,9037248-2,brian,wodlinger,Pittsburgh,PA,US,NaN
6,9981048,2018,9981048-1,ira,fox,Pittsburgh,PA,US,jacob
7,9981048,2018,9981048-3,taichiro,nishikawa,Pittsburgh,PA,US,n


Now we are done with the inital data prep work for the inventor file. Please keep in mind that we just provided some examples for you to demonstrate the process. You can add as many further steps to it as necessary. 

#### Clean Grant Data
Now we will clean and preprocess the grants data. We need a dataset that has the first, middle and last name of the researcher, the city, state, and country information. 

In [73]:
pi.head()

,PROJECT_ID,CONTACT_PI_PROJECT_LEADER,OTHER_PIS,ORGANIZATION_CITY,ORGANIZATION_STATE,ORGANIZATION_COUNTRY
0,121219,"POIRER, DAVID",NaN,TUCSON,AZ,UNITED STATES
1,121220,"HEYWARD, ANN O",NaN,BROOK PARK,OH,UNITED STATES
2,121221,"KALOGERA, VASSILIKI","LUIJTEN, ERIK",EVANSTON,IL,UNITED STATES
3,121222,"GUILDERSON, THOMAS P",NaN,SANTA CRUZ,CA,UNITED STATES
4,121223,"HAYES, GILLIAN","TENTORI, MONICA",IRVINE,CA,UNITED STATES


Here you can start cleaning the data and extract the information that we need to compare it to the inventor file. You want to find as many variables as you can match to as possible. You can also see that we have two field with names. You have the PI field, but you can extract more names from the OTHER PIDS fields if you are interested. 

In [74]:
pi[' CONTACT_PI_PROJECT_LEADER'] = clean(pi[' CONTACT_PI_PROJECT_LEADER'])

In [75]:
pi['name_first'] = pi[' CONTACT_PI_PROJECT_LEADER'].str.split(' ').str.get(1)
pi['name_last'] = pi[' CONTACT_PI_PROJECT_LEADER'].str.split(' ').str.get(0)
pi['name_middle'] = pi[' CONTACT_PI_PROJECT_LEADER'].str.split(' ').str.get(2)

In [115]:
pi[' CONTACT_PI_PROJECT_LEADER'] = pi[' CONTACT_PI_PROJECT_LEADER'].str.replace(' ','')
pi.head()

,PROJECT_ID,CONTACT_PI_PROJECT_LEADER,OTHER_PIS,ORGANIZATION_CITY,ORGANIZATION_STATE,ORGANIZATION_COUNTRY,name_first,name_last,name_middle,phonetic_first,phonetic_last,phonetic_middle
0,121219,poirerdavid,NaN,TUCSON,AZ,UNITED STATES,david,poirer,NaN,DAVAD,PARAR,NaN
1,121220,heywardanno,NaN,BROOK PARK,OH,UNITED STATES,ann,heyward,o,AN,HAYWAD,O
2,121221,kalogeravassiliki,"LUIJTEN, ERIK",EVANSTON,IL,UNITED STATES,vassiliki,kalogera,NaN,VASALAC,CALAGAR,NaN
3,121222,guildersonthomasp,NaN,SANTA CRUZ,CA,UNITED STATES,thomas,guilderson,p,TAN,GALDARSAN,P
4,121223,hayesgillian,"TENTORI, MONICA",IRVINE,CA,UNITED STATES,gillian,hayes,NaN,GALAN,HAY,NaN


We will add one more thing: **phonetic processing**.

Sometimes, words or names are recorded differently because they are written down as they sound. This can result in failed matches, because the same organization or individual will technically have different written names, even though the names would sound identically when pronounced out loud. To avoid these issues, we will add onoe more thing: a soundex (a phonetic algorithm for indexing names by sound, as pronounced in English).

The `phonetic()` function is used to convert strings into their corresponding phonetic codes. This is particularly useful when comparing names where different possible spellings make it difficult to find exact matches (e.g. Jillian and Gillian).

Let's add a column called `phonetic_name` to our existing table, which will contain the result of applying a `phonetic` function to the name (the phonetic transcription of the name). We are using a method called NYSIIS - the New York State Identification and Intelligence System phonetic code.

In [260]:
inventors["phonetic_first"] = phonetic(inventors["name_first"], method="nysiis")
inventors["phonetic_last"] = phonetic(inventors["name_last"], method="nysiis")
inventors["phonetic_middle"] = phonetic(inventors["name_middle"], method="nysiis")

pi["phonetic_first"] = phonetic(pi["name_first"], method="nysiis")
pi["phonetic_last"] = phonetic(pi["name_last"], method="nysiis")
pi["phonetic_middle"] = phonetic(pi["name_middle"], method="nysiis")

In [114]:
inventors.head(22)

,patent_id,year,inventor_id,name_first,name_last,inventor_city,inventor_state,inventor_country,name_middle,phonetic_first,phonetic_last,phonetic_middle
3,9980641,2018,5321501-6,joel,schuman,Pittsburgh,PA,US,s,JAL,SANAN,S
4,9980641,2018,9980641-3,lawrence,kagemann,Pittsburgh,PA,US,edward,LARANC,CAGANAN,EDWAD
5,9980645,2018,9037248-2,brian,wodlinger,Pittsburgh,PA,US,NaN,BRAN,WADLANGAR,NaN
6,9981048,2018,9981048-1,ira,fox,Pittsburgh,PA,US,jacob,IR,FAX,JACAB
7,9981048,2018,9981048-3,taichiro,nishikawa,Pittsburgh,PA,US,n,TACAR,NASAC,N
8,9981048,2018,9981048-4,alejandro,soto gutierrez,Pittsburgh,PA,US,NaN,ALAJANDR,SATAGATAR,NaN
9,9981915,2018,5156079-2,wei,li,Pittsburgh,PA,US,NaN,W,L,NaN
10,9981961,2018,5979076-1,zheng,li,Pittsburgh,PA,US,NaN,ZANG,L,NaN
11,9982070,2018,5077078-2,krzysztof,matyjaszewski,Pittsburgh,PA,US,a,CRSYSTAF,MATYJASAESC,A
12,9982070,2018,9982070-2,pawel,chmielarz,Pittsburgh,PA,US,NaN,PAL,CNALAR,NaN


In [138]:
pi.head()

,PROJECT_ID,CONTACT_PI_PROJECT_LEADER,OTHER_PIS,ORGANIZATION_CITY,ORGANIZATION_STATE,ORGANIZATION_COUNTRY,name_first,name_last,name_middle,phonetic_first,phonetic_last,phonetic_middle
0,121219,poirerdavid,NaN,TUCSON,AZ,UNITED STATES,david,poirer,NaN,DAVAD,PARAR,NaN
1,121220,heywardanno,NaN,BROOK PARK,OH,UNITED STATES,ann,heyward,o,AN,HAYWAD,O
2,121221,kalogeravassiliki,"LUIJTEN, ERIK",EVANSTON,IL,UNITED STATES,vassiliki,kalogera,NaN,VASALAC,CALAGAR,NaN
3,121222,guildersonthomasp,NaN,SANTA CRUZ,CA,UNITED STATES,thomas,guilderson,p,TAN,GALDARSAN,P
4,121223,hayesgillian,"TENTORI, MONICA",IRVINE,CA,UNITED STATES,gillian,hayes,NaN,GALAN,HAY,NaN


<span style="color:red">**Checkpoint 2: Clean organization-level data**</span>

Use `clean` function above, remove white space, and find phonetic transcriptions of organization names.

In [195]:
# your code here...
#organizations.head()
organizations[' ORGANIZATION_NAME'] = clean(organizations[' ORGANIZATION_NAME'])
organizations[' ORGANIZATION_NAME'] = organizations[' ORGANIZATION_NAME'].str.replace(' ','')

#organizations[' ORGANIZATION_NAME']
organizations[" PHONETIC_ORGANIZATION_NAME"] = phonetic(organizations[" ORGANIZATION_NAME"], method="nysiis")
#organizations.head()

pi_organization = pi.merge(organizations, left_on='PROJECT_ID', right_on='PROJECT_ID', how='left')

print(organizations.shape)
print(pi.shape)
print(pi_organization.shape)

pi_organization.head()

(109746, 6)
(109746, 12)
(109746, 17)


,PROJECT_ID,CONTACT_PI_PROJECT_LEADER,OTHER_PIS,ORGANIZATION_CITY_x,ORGANIZATION_STATE_x,ORGANIZATION_COUNTRY_x,name_first,name_last,name_middle,phonetic_first,phonetic_last,phonetic_middle,ORGANIZATION_NAME,ORGANIZATION_CITY_y,ORGANIZATION_STATE_y,ORGANIZATION_COUNTRY_y,PHONETIC_ORGANIZATION_NAME
0,121219,poirerdavid,NaN,TUCSON,AZ,UNITED STATES,david,poirer,NaN,DAVAD,PARAR,NaN,universityofarizona,TUCSON,AZ,UNITED STATES,UNAVARSATYAFARASAN
1,121220,heywardanno,NaN,BROOK PARK,OH,UNITED STATES,ann,heyward,o,AN,HAYWAD,O,ohioaerospaceinstitute,BROOK PARK,OH,UNITED STATES,OHARASPACANSTATAT
2,121221,kalogeravassiliki,"LUIJTEN, ERIK",EVANSTON,IL,UNITED STATES,vassiliki,kalogera,NaN,VASALAC,CALAGAR,NaN,northwesternuniversity,EVANSTON,IL,UNITED STATES,NARTWASTARNANAVARSATY
3,121222,guildersonthomasp,NaN,SANTA CRUZ,CA,UNITED STATES,thomas,guilderson,p,TAN,GALDARSAN,P,universityofcaliforniasantacruz,SANTA CRUZ,CA,UNITED STATES,UNAVARSATYAFCALAFARNASANTACR
4,121223,hayesgillian,"TENTORI, MONICA",IRVINE,CA,UNITED STATES,gillian,hayes,NaN,GALAN,HAY,NaN,universityofcaliforniairvine,IRVINE,CA,UNITED STATES,UNAVARSATYAFCALAFARNARVAN


In [216]:
#print(assignees.shape)
print(inventors.groupby(["patent_id","inventor_id","inventor_city"]).ngroups)

inventors =  inventors.drop_duplicates(subset=["patent_id","inventor_id","inventor_city"])



1833


In [250]:
print(assignees.groupby(["patent_id", "organization", "org_city", "org_state", "org_country"]).ngroups)
assignees = assignees.drop_duplicates(subset=["patent_id", "organization", "org_city", "org_state", "org_country"])

assignees_merged = assignees.merge(organizations, left_on="organization", right_on=" ORGANIZATION_NAME", how='left')
print(assignees_merged.shape)

3744
(383222, 11)


In [253]:
assignees
#organizations

,patent_id,organization,org_city,org_state,org_country
0,9992413,raytheoncompany,Lexington,MA,US
1,RE45718,bostonscientificcorporation,Lexington,MA,US
2,9992413,raytheoncompany,Waltham,MA,US
3,9988932,honeywellinternationalinc,Waltham,MA,US
4,9991578,honeywellinternationalinc,Waltham,MA,US
5,9989730,honeywellinternationalinc,Waltham,MA,US
6,9993577,trusteesofbostonuniversity,Waltham,MA,US
7,9986740,brandeisuniversity,Waltham,MA,US
8,9982790,bostondynamicsinc,Waltham,MA,US
9,9992413,raytheoncompany,Los Angeles,CA,US


In [146]:
#Change assignees, they will be merged with organizations through organization:
assignees['organization'] = clean(assignees['organization'])
assignees['organization'] = assignees['organization'].str.replace(' ','')

assignees["phonetic_organization"] = phonetic(assignees[" organization"], method="nysiis")

# pi will be merged with organizations via project_id\


KeyError: ' organization'

# Deterministic Matching

### Individual-level links

Let's first do an exact match (merge) on the **first, middle, last name** and examine the results.

In [ ]:
# Merge on first, middle, last name
merge_name = inventors.merge(pi,on=['name_first','name_last','name_middle'])

Let's count how many unique inventor IDs and PI names got matched.

In [ ]:
# Let's check the count of unique inventor_ids
merge_name['inventor_id'].nunique()

We can now merge on the **phonetic first, middle, last name**.

In [ ]:
# Merge on phonetic first, middle, last name
merge_name_phonetic_middle = inventors.merge(pi,on=['phonetic_first','phonetic_last','phonetic_middle'])

In [ ]:
# Let's check the count of unique inventor_ids
merge_name_phonetic_middle['inventor_id'].nunique()

You can enlarge your links and only merge on **first and last name**.

In [ ]:
merge_name_first_last = inventors.merge(pi,on=['name_first','name_last'])

In [ ]:
# Check the counts
merge_name_first_last['inventor_id'].nunique()

You can also try merging on **phonetic first and last name**.

In [ ]:
merge_name_phonetic = inventors.merge(pi,on=['phonetic_first','phonetic_last'])

In [ ]:
merge_name_phonetic['inventor_id'].nunique()

What if we blocked by state?

In [ ]:
# Rename columns in both datasets to the same name
inventors = inventors.rename(columns={'inventor_state':'state'})
pi = pi.rename(columns={' ORGANIZATION_STATE':'state'})

Merge on **first, middle, last name, and state**.

In [ ]:
merge_name_state = inventors.merge(pi,on=['name_first','name_last','name_middle','state'])

How many matches did we get?

In [ ]:
merge_name_state['inventor_id'].nunique()

In [ ]:
merge_phonetic_name_state = inventors.merge(pi,on=['phonetic_first','phonetic_last','phonetic_middle','state'])

In [ ]:
merge_phonetic_name_state['inventor_id'].nunique()

Now let's try blocking by **full name, city and state**.

In [ ]:
inventors.head()

In [ ]:
pi.head()

In [ ]:
inventors['inventor_city'] = inventors['inventor_city'].str.upper()

In [ ]:
inventors = inventors.rename(columns={'inventor_city':'city'})

In [ ]:
pi = pi.rename(columns={' ORGANIZATION_CITY':'city'})

In [ ]:
merge_name_location_city = inventors.merge(pi, on=['name_first','name_last','name_middle','state','city'])

How many matches did we get?

In [ ]:
merge_name_location_city['inventor_id'].nunique()

In [ ]:
merge_phonetic_name_location_city = inventors.merge(pi, on=['phonetic_first','phonetic_last','state','city'])

In [ ]:
merge_phonetic_name_location_city['inventor_id'].nunique()

<span style="color:red">**Checkpoint 3: Deterministic matching of organization names**</span>

Now try doing different types of merges using organization names, blocking on different variables and seeing the differences.

In [ ]:
# your code here...



**Optional: Regular Expressions - regex**

When defining a regular expression search pattern, it is a good idea to start out by writing down, explicitly, in plain English, what you are trying to search for and exactly how you identify when you've found a match.
For example, if we look at an author field formatted as "<last_name> , <first_name> <middle_name>", in plain English, this is how I would explain where to find the last name: "starting from the beginning of the line, take all the characters until you see a comma."

We can build a regular expression that captures this idea from the following components:
- ^ Matches beginning of the line
- . Matches any character
- .+ A modifier that means "match one or more of the preceding expression"

In a regular expression, there are special reserved characters and character classes like those in the list above. Anything that is not a special character or class is just looked for explicitly (for example, a comma is not a special character in regular expressions, so if it is in a regular expression pattern, the regular expression processor will just be looking for a comma in the string, at that point in the pattern).

Note: if you want to actually look for one of these reserved characters, it must be escaped, so that, for example, the expression looks for a literal period, rather than the special regular expression meaning of a period. To escape a reserved character in a regular expression, precede it with a back slash ( "." ).
This results in the regular expression: ^.+,

We start at the beginning of the line ( "^" ), matching any characters ( ".+" ) until we come to the literal character of a comma ( "," ).

In python, to use a regular expression like this to search for matches in a given string, we use the built-in "re" package ( https://docs.python.org/2/library/re.html ), specifically the "re.search()" method. To use "re.search()", pass it first the regular expression you want to use to search, enclosed in quotation marks, and then the string you want to search within. 

#### REGEX CHEATSHEET


    - abc...     Letters
    - 123...     Digits
    - \d         Any Digit
    - \D         Any non-Digit Character
    - .          Any Character
    - \.         Period
    - [a,b,c]    Only a, b or c
    - [^a,b,c]   Not a,b, or c
    - [a-z]      Characters a to z
    - [0-9]      Numbers 0 to 9
    - \w any     Alphanumeric chracter
    - \W         any non-Alphanumeric character
    - {m}        m Repetitions
    - {m,n}      m to n repetitions
    - *          Zero or more repetitions
    - +          One or more repetitions
    - ?          Optional Character
    - \s         any Whitespace
    - \S         any non-Whitespace character
    - ^...$      Starts & Ends
    - (...)      Capture Group
    - (a(bc))    Capture sub-Group
    - (.*)       Capture All
    - (abc|def)  Capture abc or def
     
#### EXAMPLES
    - (\d\d|\D) will match 22X, 23G, 56H, etc...
    - \w will match any characters between 0-9 or a-z
    - \w{1-3} will match any alphanumeric character of a length of 1 to 3. 
    - (spell|spells) will match spell or spells

## References and Further Readings


### Record Linkage
Winkler, William E. 2009. “Record Linkage.” In Handbook of Statistics 29a, Sample Surveys: Design, Methods and Applications, edited by D. Pfeffermann and C.
R. Rao, 351–80. Elsevier.

Lane, Julia, Ian Foster, Rayid Ghani, Ron S. Jarmin, Frauke Kreuter (editors), Big Data and Social Science: A Practical Guide to Methods and Tools, Chapman and Hall/CRC Press, 2016. 
https://coleridge-initiative.github.io/big-data-and-social-science/chap-link.html

### Parsing

* Python online documentation: https://docs.python.org/2/library/string.html#deprecated-string-functions
* Python 2.7 Tutorial (Splitting and Joining Strings): http://www.pitt.edu/~naraehan/python2/split_join.html

### Regular Expression

* Python documentation: https://docs.python.org/2/library/re.html#regular-expression-syntax
* Online regular expression tester (good for learning): http://regex101.com/

### Record Linkage Python package
* `recordlinkage` Python package: https://recordlinkage.readthedocs.io/en/latest/index.html